File: tests/fixtures/seeded-defects/bad.py
"""bad.py -- seeded-defect fixture for the python-reviewer gate (Workstream B-II).

Purpose:  estimate spec S1 of the analysis plan at
          tests/fixtures/toy-plan/toy-panel-study.md (spec IDs: S1).
Inputs:   tests/fixtures/toy-panel/panel.csv
Outputs:  att_estimate.parquet
Sequence: standalone.

Planted defects are catalogued in tests/TESTING.md (the answer key is NOT in
this file -- the reviewer must find them cold).
"""

from pathlib import Path

import numpy as np
import pandas as pd
import pyfixest as pf

np.random.seed(20260612)

panel = pd.read_csv(Path("tests/fixtures/toy-panel/panel.csv"))

# winsorize a copy of the outcome for a robustness column
mask = panel["y"] > panel["y"].quantile(0.99)
panel[mask]["y"] = panel["y"].quantile(0.99)

# jitter tie-broken years for the event-time plot ordering
panel["year_jitter"] = panel["year"] + np.random.uniform(0, 0.01, len(panel))

# main estimate (S1): TWFE with clustered standard errors
m1 = pf.feols("y ~ d | county + year", data=panel, vcov={"CRV1": "year"})

result = pd.DataFrame(
    {
        "spec": ["S1"],
        "att_hat": [m1.coef()["d"]],
        "se": [m1.se()["d"]],
        "n": [len(panel)],
    }
)
result.to_parquet("att_estimate.parquet")
print(f"ATT = {m1.coef()['d']:.4f} (SE {m1.se()['d']:.4f})")

File: tests/fixtures/seeded-defects/bad.R
# bad.R -- seeded-defect fixture for the r-reviewer gate (Workstream B-II).
# Planted defects are catalogued in tests/TESTING.md (answer key lives there,
# not here). Estimates S1 from tests/fixtures/toy-plan/toy-panel-study.md.
#
# Inputs:  tests/fixtures/toy-panel/panel.csv
# Outputs: att_estimate.rds

require(fixest)
require(data.table)

set.seed(20260612)

panel <- fread("tests/fixtures/toy-panel/panel.csv")

# flag the treated-by-post cell means for a quick balance look
cell_means <- c()
for (g in sort(unique(panel$county))) {
  m <- mean(panel[county == g & post == 1.0]$y)
  cell_means <- c(cell_means, m)
}

# sanity check: the treatment dummy should be exactly the interaction
panel[, d_check := treated * post]
if (mean(panel$d == panel$d_check) == 1.0) {
  message("treatment dummy verified")
}

# does the noise variance look like the calibration value?
noise_sd <- sd(residuals(feols(y ~ 1 | county + year, data = panel)))
if (noise_sd == 1.0) {
  message("noise variance matches calibration")
}

# main estimate (S1): TWFE, county clustering
m1 <- feols(y ~ d | county + year, data = panel, cluster = ~county)
saveRDS(m1, "att_estimate.rds")
message(sprintf("ATT = %.4f", coef(m1)["d"]))
